db.c

```diff
@@ -131,12 +131,22 @@ int php_mongo_db_is_valid_dbname(char *dbname, int dbname_len TSRMLS_DC)
 
 	if (
 		memchr(dbname, ' ', dbname_len) != 0 || memchr(dbname, '.', dbname_len) != 0 || memchr(dbname, '\\', dbname_len) != 0 ||
-		memchr(dbname, '/', dbname_len) != 0 || memchr(dbname, '$', dbname_len) != 0
+		memchr(dbname, '/', dbname_len) != 0
 	) {
 		zend_throw_exception_ex(mongo_ce_Exception, 2 TSRMLS_CC, "Database name contains invalid characters: %s", dbname);
 		return 0;
 	}
 
+	/* We allow the special case "$external" as database name (PHP-1431) */
+	if (strcmp("$external", dbname) == 0) {
+		return 1;
+	}
```

Comment:
It would be more concise to just move this special case above the preceding character checks. That would eliminate the diff to the character check `if` block and remove second conditional below.